core/dbt/graph/selector_methods.py

```diff
@@ -577,6 +578,39 @@ def search(
             if node in matches:
                 yield node
 
+class SourceFreshSelectorMethod(SelectorMethod): #TODO: this requires SelectorMethod to have current_state as an argument. currently, this works because it's all hard-coded
+    def search(
+        self, included_nodes: Set[UniqueId], selector: str
+    ) -> Iterator[UniqueId]:
+
+        if self.previous_state is None or \
+           self.previous_state.current_sources is None or \
+           self.previous_state.archive_sources is None:
+            raise InternalException(
+                'No previous state comparison freshness results in sources.json'
+            )
+        
+        current_state_sources = {
+            result.unique_id:result.max_loaded_at for result in self.previous_state.current_sources.results
+        }
+
+        archive_state_sources = {
+            result.unique_id:result.max_loaded_at for result in self.previous_state.archive_sources.results
+        }
+
+        matches = set()
+        for unique_id in current_state_sources:
+            if unique_id not in archive_state_sources:
+                matches.add(unique_id)
+            elif selector == 'yes' and current_state_sources.get(unique_id) > archive_state_sources.get(unique_id):
+                matches.add(unique_id)
+            elif selector == 'no' and current_state_sources.get(unique_id) <= archive_state_sources.get(unique_id):
+                matches.add(unique_id)
```

Comment:
Ah - I now remember discussing the case where someone does _not_ want to build models downstream of a source, unless all tables in that source are fresh, so as to avoid uncoordinated updates.

Then, the inclusion of `source_fresh:no` exists to support:
```
dbt build -s source_fresh:yes --exclude source_fresh:no
```

Even then, I think this is likely to break down. Suppose we have three runs (A, B, C), each of which checks source freshness and defers to the previous run. If half of the sources update between A &rarr; B, and half of the sources update between B &rarr; C — we don't end up updating any downstream models, because in neither comparison (A &rarr; B, B &rarr; C) do _all_ source tables appear "fresher" than they were before.

To actually solve for that aim, I think we'd need to coordinate source tables with previous model build times. ("Have all source tables upstream of model X been updated since model X last run?") That information would be available from the previous `run_results.json` — but now we're talking about even more moving pieces, plus factoring in graph traversal (?).

At this level of strictness, someone would be better off triggering `dbt build source:specific_src+` when (and only when) they've confirmed that `specific_src` has finished updating all tables.

Reply:
Agreed on removing the complex orchestration example from scope.

We will change the flag to something like: `source_fresher`

Example: `dbt build --select source_fresher`